### PR TITLE
feat!: `worker.dev_create_account`, `worker.dev_deploy` create subaccounts of root account instead of tla (sandbox, testnet)

### PR DESCRIPTION
**Pull request by AlexKushnir1**
Renamed method (create_tla) and added .testnet to the randomly generated ID sent to the helper when creating a dev account

`worker.dev_create_account` has now become `worker.dev_create_tla`, `worker.dev_deploy` has now become `worker.dev_deploy_tla`

___

- [x] main branch (base branch)
  - [x] sandbox
    - [x] creates top level account (`worker.create_tla`, `worker.create_tla_and_deploy` ,`worker.dev_create_account`, `worker.dev_deploy`)
    ```bash
    [src/main.rs:18:9] &result = Account {
        id: AccountId(
            "dev-20241224213704-37976876532241",
        ),
    }
    balance = 100.00 NEAR
    ```
  - [x] testnet
    - [x] ERRORs on top level account creation 

    ```bash
    Error: Execution

    Caused by:
        Action #0: A top-level account ID AccountId("dev-20241224215752-62591402306687") can't be created
    by AccountId("testnet"), short top-level account IDs can only be created by AccountId("registrar")
    ```

- [x] remove-using-tla-on-testnet branch (head branch)
  - [x] sandbox
    - [x] creates top level account (`worker.create_tla`, `worker.create_tla_and_deploy`, `worker.dev_create_tla`, `dev_deploy_tla`)

    ```bash
    [src/main.rs:19:9] &result = Account {
        id: AccountId(
            "dev-20241224222905-84560672890155",
        ),
    }
    balance = 100.00 NEAR
    ```
    - [x] creates root account sub-account (`worker.create_root_account_subaccount`,` worker.create_root_account_subaccount_and_deploy`,` worker.dev_create_account`,` worker.dev_deploy`)

    ```bash
    [src/main.rs:35:13] &root_account = Account {
        id: AccountId(
            "test.near",
        ),
    }
    balance = 1000000000.00 NEAR
    [src/main.rs:48:9] &result = Account {
        id: AccountId(
            "dev-20241225121559-79849516817890.test.near",
        ),
    }
    balance = 100.00 NEAR
    ```
  - [x] testnet
    - [x] FAILS to compile on top level account creation (no such method) 
    - [x] creates root account sub-account (`worker.create_root_account_subaccount`,` worker.create_root_account_subaccount_and_deploy`,` worker.dev_create_account`,` worker.dev_deploy`)

    ```bash
    [src/main.rs:62:13] &root_account_id = AccountId(
        "testnet",
    )
    balance = 15561968.34 NEAR
    [src/main.rs:74:9] &result = Account {
        id: AccountId(
            "dev-20241225121906-39307469198620.testnet",
        ),
    }
    balance = 10.01 NEAR
    ```

### CODE REVIEW

**Comment by AlexKushnir1** [comment]
@race-of-sloths

**Comment by race-of-sloths** [comment]
@AlexKushnir1 Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!


[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/AlexKushnir1?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/AlexKushnir1?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/AlexKushnir1)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/AlexKushnir1)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: workspaces/tests/account.rs
Comment (frol) [inline workspaces/tests/account.rs:11]: Testing with testnet will drain the faucet too fast, unless we delete the accounts and return all unused tokens back to the faucet #215.

Either resolve #215 (at least manually deleting the accounts in the test) or revert it back to `sandbox`

File: workspaces/src/network/variants.rs
Comment (frol) [inline workspaces/src/network/variants.rs:19]: After a second thought, I think it is better to have two traits:

1. TopLevelAccountCreator (it is still possible to do that in the sandbox as we control the registrar account, so there is still a value to have this trait as it was before the PR)
2. SponsoredAccountCreator

Sandbox can implement both, testnet will implement only SponsoredAccountCreator, and custom wrappers can even implement it for mainnet.

File: workspaces/src/network/testnet.rs
Comment (frol) [inline workspaces/src/network/testnet.rs:82]: Let's make the `id` argument (`create_dev_account`/`create_sponsored_account`'s function argument) explicitly named as  `subaccount_prefix` or something along those lines, and document that on the trait - it is expected that this name does not contain `.` (`testnet` cannot create `sub.sub.testnet` accounts).

Comment (frol) [inline workspaces/src/network/testnet.rs:82]: I suggest to avoid unnecessary hardcodings:

```suggestion
        let id = AccountId::from_str(format!("{}.{}", id, self.into().root_id)).map_err(...)?;
```

**Review by frol — CHANGES_REQUESTED** [review]

File: workspaces/src/network/variants.rs
Comment (AlexKushnir1) [reply workspaces/src/network/variants.rs]: Spent time thinking and did not understand how to implement functions(https://github.com/near/near-workspaces-rs/pull/369/files#diff-2d190d038b29b514c04b33ac855123d6949ed0f3e8b5b8b7034b0de327992e79R62) , that must to get both tla and sponsored accounts. Also as the sandbox can implement both traits, how can we choose method that will be used

File: workspaces/tests/account.rs
Comment (frol) [inline workspaces/tests/account.rs:44]: Failing tests must clean up the accounts as well to avoid draining the faucet account.

Just for context, the previous comment on this topic was here: https://github.com/near/near-workspaces-rs/pull/369/files#r1713257285, please, either fix the drainage or go back to sandbox.

File: workspaces/src/network/variants.rs
Comment (frol) [inline workspaces/src/network/variants.rs:132]: Do we still need `AllowDevAccountCreation`? I would implement `SponsortedAccountCreator` for testnet and sandbox networks (the implementations will be different, but the public interface can be the same) and call `DevNetwork = Network + SponsoredAccountCreator`, and then here it will be just `T: DevNetwork + 'static`

File: workspaces/src/network/variants.rs
Comment (frol) [inline workspaces/src/network/variants.rs:133]: Having two methods with a similar name (`create_dev_account`, `dev_create_account`) is really confusing. I'd resolve this naming issue and feel free to make the breaking change here since we will release it as a breaking release anyway.

Ugly naming ideas:

* `create_sponsored_account(prefix)` + `create_sponsored_account_with_random_name()`
* `create_sponsored_account(prefix)` + `create_randomly_named_sponsored_account()`
* `create_sponsored_account(prefix)` + `create_dev_account()` (`dev_account` implies that we don't care about the specific name it gets, so `dev-123123123123` is fine)

**Review by frol — DISMISSED** [review]
@AlexKushnir1 Great job! We are getting very close to solving it once and for all!

@race-of-sloths score 13

File: workspaces/tests/account.rs
Comment (AlexKushnir1) [reply workspaces/tests/account.rs]: Returned to the sandbox. The main idea of ​​staying on testnet was that there are no tests on testnet

**Comment by race-of-sloths** [comment]
### ❌ PR inactive
This PR has been inactive for two weeks. It's now marked as stale. 
>[!TIP]
If you think it's a mistake, you can restart the bot with `include` command to proceed 🔄

File: workspaces/tests/account.rs
Comment (akorchyn) [reply workspaces/tests/account.rs:11]: I think this is still actual.
Even though, we don't use testnet in tests. I think we should motivate users to delete and remove created accounts.

Maybe, we should have something as `Drop` implementation that would `panic` if user is not deleted. 
Just thinking loud :)

File: workspaces/src/network/variants.rs
Comment (akorchyn) [inline workspaces/src/network/variants.rs:97]: should we rename it to generate_sponsored_credentials?

File: workspaces/src/network/variants.rs
Comment (akorchyn) [inline workspaces/src/network/variants.rs:175]: `dev_generate_credentials`
would be easier to find

File: workspaces/tests/account.rs
Comment (akorchyn) [inline workspaces/tests/account.rs:35]: What's that file?

**Review by akorchyn — COMMENTED** [review]
Overall, looks good to me. I would want to have a one more review here from
@dj8yfo or @PolyProgrammist

File: .github/workflows/test.yml
Comment (AlexKushnir1) [reply .github/workflows/test.yml:50]: No reasons for prod, had a error while checking on macOS and expected to run ci with macOS firslty. Fixed it, thanks!

File: workspaces/tests/account.rs
Comment (AlexKushnir1) [reply workspaces/tests/account.rs:35]: This file is created on previous lines. I`m realy not sure it makes much sense since it`s stored in ./target but it seems like one less unnecessary file

File: workspaces/tests/account.rs
Comment (AlexKushnir1) [reply workspaces/tests/account.rs:12]: Yeah, that makes more sense. Do you think the `dev_deploy` is intuitive enough?

File: workspaces/src/network/variants.rs
Comment (AlexKushnir1) [reply workspaces/src/network/variants.rs:97]: It will probably get confusing inasmuch as we have the following two unrelated to this function  `create_sponsored_account` and `create_sponsored_account_and_deploy`

File: workspaces/src/network/variants.rs
Comment (AlexKushnir1) [reply workspaces/src/network/variants.rs:175]: Guess we can change it to `generate_sponsored_credentials` since we use it to create sponsored accounts. WDYT?

File: workspaces/tests/account.rs
Comment (akorchyn) [reply workspaces/tests/account.rs:12]: Well, I think, people get used to it already. So it's more to keep things similar :)

File: workspaces/src/network/variants.rs
Comment (akorchyn) [reply workspaces/src/network/variants.rs:97]: fair

**Comment by dj8yfo** [comment]
will be adding some minor changes here

File: workspaces/tests/account.rs
Comment (dj8yfo) [reply workspaces/tests/account.rs]: i don't think this is a real issue of users, who just want to routinely do their integration tests and get some test NEAR in the process. 
this is an issue of users, who intentionally want to drain faucet just for fun or for some other reason, and they still can do it via other means, outside of `near-workspaces-rs` api.

File: workspaces/src/network/testnet.rs
Comment (dj8yfo) [inline workspaces/src/network/testnet.rs:124]: this appears to never haver been run at least once, as it didn't work

File: workspaces/src/network/testnet.rs
Comment (dj8yfo) [inline workspaces/src/network/testnet.rs:117]: this appears to never have been run, as it didn't work

**Review by dj8yfo — APPROVED** [review]

File: workspaces/tests/account.rs
Comment (akorchyn) [reply workspaces/tests/account.rs]: Well, if some test runs in CI and creates an account with Near every run. 100 runs === 1000 NEAR drained.

Comment (AlexKushnir1) [reply workspaces/tests/account.rs]: Just thinking out loud. I see this as some kind of enum parameter that specifies how long the user will use this account and delete it after that period. If user isn"t intrested in this staff we can set default value such as s short lifetime

**Pull request merged**
